### PR TITLE
Listener Support Implementation

### PR DESCRIPTION
**Pull request by MadhurangaWije**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -51,3 +64,64 @@ WhatsApp messaging requires users to use predefined [message templates](https://
...
+    - Callback webhook URL has to be registered at the time of making the call (from the client connector)
+
+### Getting started
+
+1.  Refer the [Get Started](https://ballerina.io/v1-1/learn/) section to download and install Ballerina.
+
+2. Import the Twilio Webhook module to your Ballerina program as follows.
+
+	```ballerina
+	import ballerina/io;
```
Comment (kasun04) [inline README.md:98]: formatting error?

File: modules/webhook/tests/test.bal [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+import ballerina/config;
+import ballerina/test;
+import ballerina/websub;
+import ballerinax/twilio;
+import ballerina/io;
+import ballerina/runtime;
+import ballerina/log;
+
+listener Listener twilioListener = new (8080);
```
Comment (kasun04) [inline modules/webhook/tests/test.bal:25]: Should this be a hardcoded port?

File: modules/webhook/tests/test.bal [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+// Mock service for testing webhook events
+
+@websub:SubscriberServiceConfig {subscribeOnStartUp: false}
+service websub:SubscriberService /twilio on twilioListener {
+    remote function onNotification(websub:Notification notification) {
+
+        io:println("");
+        io:println(notification.getFormParams().toBalString());
+        io:println("");
```
Comment (kasun04) [inline modules/webhook/tests/test.bal:45]: Why we have system io in this code? Can't we use log interfaces?

File: modules/webhook/twilio_webhook.bal [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+import ballerina/websub;
+
+////////////////////////////////////////////////////////////////////
+/// Twilio Webhook Listener (WebSub Subscriber Service Listener) ///
+////////////////////////////////////////////////////////////////////
+# Object representing the Twilio Webhook (WebSub Subscriber Service) Listener.
+public class Listener {
```
Comment (kasun04) [inline modules/webhook/twilio_webhook.bal:22]: Should we name it 'TwilioWebhookListener'?

File: twilio_endpoint.bal [hunk 5]
```diff
@@ -121,14 +128,30 @@ public client class Client {
...
         requestBody = check createUrlEncodedRequestBody(requestBody, FROM, fromNo);
         requestBody = check createUrlEncodedRequestBody(requestBody, TO, toNo);
         requestBody = check createUrlEncodedRequestBody(requestBody, URL, twiml);
+
+        if (statusCallback is StatusCallback) {
+            requestBody = check createUrlEncodedRequestBody(requestBody, STATUS_CALLBACK_URL, statusCallback.url);
+            requestBody = check createUrlEncodedRequestBody(requestBody, STATUS_CALLBACK_METHOD, statusCallback.method);
+
+            string[]|() events = statusCallback?.events;
+
```
Comment (kasun04) [inline twilio_endpoint.bal:147]: Avoid redundant newlines.

**Review by kasun04 — CHANGES_REQUESTED** [review]

File: modules/webhook/tests/test.bal [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+boolean smsQueuedNotified = false;
+boolean smsSentNotified = false;
+boolean smsRecievedNotified = false;
+
+boolean callQueuedNotified = false;
+boolean callRingingNotified = false;
+boolean callInProgressNotified = false;
+boolean callCompletedNotified = false;
+
+
```
Comment (LakshanSS) [inline modules/webhook/tests/test.bal:36]: Remove unwanted newline

File: modules/webhook/tests/test.bal [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            if (payload.CallStatus === QUEUED) {
+                callQueuedNotified = true;
+            } else if (payload.CallStatus === RINGING) {
+                callRingingNotified = true;
+            } else if (payload.CallStatus === IN_PROGRESS) {
+                callInProgressNotified = true;
+            } else if (payload.CallStatus === COMPLETED) {
+                callCompletedNotified = true;
+            }
+
```
Comment (LakshanSS) [inline modules/webhook/tests/test.bal:72]: Remove unwanted newline

File: modules/webhook/tests/test.bal [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        }
+    }
+
+}
+
+
+
```
Comment (LakshanSS) [inline modules/webhook/tests/test.bal:79]: Check all the places and remove unwanted newlines

File: modules/webhook/twilio_webhook.bal [hunk 9]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (LakshanSS) [inline modules/webhook/twilio_webhook.bal:1]: Year has to be changed to 2021

File: modules/webhook/twilio_webhook_constants.bal [hunk 10]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (LakshanSS) [inline modules/webhook/twilio_webhook_constants.bal:1]: Change the year to 2021

File: twilio_types.bal [hunk 11]
```diff
@@ -218,3 +218,10 @@ public type TwilioError distinct error;
 
 # Represents the Twilio module related error.
 public type Error TwilioError;
+
+
+public type StatusCallback record {
+    string url;
+    string method;
+    string[] events?;
+};
```
Comment (LakshanSS) [inline twilio_types.bal:227]: Add a newline in the end of the file

File: modules/webhook/tests/test.bal (hunk 3 above)
Comment (MadhurangaWije) [reply modules/webhook/tests/test.bal]: those were added for a debug purpose, and have been missed out to remove. now removed

File: README.md (hunk 1 above)
Comment (MadhurangaWije) [reply README.md:98]: Please note that this formatting error, does not exist in the file, somehow it appears here in the compare mode.

File: twilio_endpoint.bal [hunk 12]
```diff
@@ -121,14 +128,29 @@ public client class Client {
...
         requestBody = check createUrlEncodedRequestBody(requestBody, FROM, fromNo);
         requestBody = check createUrlEncodedRequestBody(requestBody, TO, toNo);
         requestBody = check createUrlEncodedRequestBody(requestBody, URL, twiml);
+
+        if (statusCallback is StatusCallback) {
+            requestBody = check createUrlEncodedRequestBody(requestBody, STATUS_CALLBACK_URL, statusCallback.url);
+            requestBody = check createUrlEncodedRequestBody(requestBody, STATUS_CALLBACK_METHOD, statusCallback.method);
+            string[]? events = statusCallback?.events;
+            
+            if(events is string[]) {
```
Comment (LakshanSS) [inline twilio_endpoint.bal:147]: Need a space after if

File: modules/webhook/tests/test.bal [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+            } else if (payload.CallStatus == RINGING) {
+                callRingingNotified = true;
+            } else if (payload.CallStatus == IN_PROGRESS) {
+                callInProgressNotified = true;
+            } else if (payload.CallStatus == COMPLETED) {
+                callCompletedNotified = true;
+            }
+        }
+    }
+
```
Comment (LakshanSS) [inline modules/webhook/tests/test.bal:74]: Remove unwanted new lines in all the similar places

**Review by kasun04 — APPROVED** [review]

**Pull request merged**